src/HtmlReader/architecture.md

```diff
@@ -0,0 +1,78 @@
+# `useHTMLReader` Architecture
+
+This file is meant as a reference to document the current state of the `useHTMLReader` hook. It is specifically aimed at developers working with this code.
+
+## Style
+
+For controlling the style of the content, we use [Readium-CSS](https://github.com/readium/readium-css), which is also what R2D2BC used before us. By simply setting css variables, it allows us to control font size, color mode, font family, and more. Thus, to change the color mode, we simply need to change a CSS variable inside the iframe.
+
+Readium-CSS also enables a "paginated" mode, where the full HTML resource is broken into multiple columns that are then rendered side by side in a horizontally scrolling container. Thus, to move to the next page in paginated mode, we simply scroll the user horizontally by one page-width.
+
+## Lifecycle and State
+
+We use `React.useReducer` as the single source of truth for state in this hook. Whenever anything changes, it should go through `useReducer`. If you render with the same state object twice, you should always get the same thing.
+
+Given that there are more than 10 properties in the state object, and each one has multiple possible values, there are seemingly infinite possible states. However, the range of _valid_ states is actually much smaller. These states are:
+
+- `InactiveState`: The hook was not provided any arguments, and thus it is inactive.
```

Comment:
Context:  So I'm starting my review/exploration with this file.  

Is it convention to call everything ThingState rather than ActiveState.FetchingResource (eg: https://github.com/NYPL-Simplified/web-reader/pull/91/files#diff-40c72fa84cd3e89b4e8825c1858a53ddb9d56a42805fcf97192b24aabbefe068R20)?

If someone were to say "oh my state is special, it's gotta be a new state" - where would they add it? How would they determine if that's actually true or not?

Reply:
That's a good question, I should think about this a little more. You're asking whether these are all a "sub-state" of the ActiveState?

In terms of how to know when you need a new one... It's a little bit arbitrary. When I first built this, I didn't have these split into multiple states. It was all just one `ReaderState`, but then I constantly had to check `if(state.isIframeLoaded && state.iframe && state.resource)` etc. So to minimize the number of those checks, I separated them into distinct states. Then we know that `if(state.state === 'READY')` then all those things are defined. It's just a convenient way to separate the complex number of options into distinct states so we can know what is defined at different times.

Reply:
> You're asking whether these are all a "sub-state" of the ActiveState?

Yes, or if they're all a "sub-state" of _something_.  clearly InactiveState isn't a substate of ActiveState.  The idea is to try to put all this in an enum that's referenced everywhere, so that, for example, we don't have a chance of doing something like `state.state === 'NOT_QUITE_READY'` without adding it to this documentation

Reply:
In essence, understanding all the states the reader can be in is the most important thing, so I'm wondering if there's a way we can define all of it in one place in the code, as well as in the documentation.  ActiveState seemed like the most obvious place to do it.

Reply:
I suppose that's what I was trying to do in the `/HtmlReader/types.ts` file where I defined all the state types. Do you think that works?